doc/testing.md

```diff
@@ -0,0 +1,133 @@
+# Testing hammer commands
+
+Commands from `hammer-cli-foreman` are tested by two types of tests:
+- *unit tests* - They're placed in `test/unit` and test only basic functionality of the commands like accepted options and field labels printed to stdout. Unit tests for commands are no longer extended.
+- *functional tests* - They're placed in `test/functional` and aim to test commands in broader context. Functional tests require mocking the api calls and should also test output based on data returned from the api. This is preferable way of testing the commands.
+
+
+## Running the tests
+
+```bash
+bundle install
+bundle exec rake test
+```
+
+Generated coverage reports are stored in ./coverage directory.
+
+There is support for testing against API documentation generated from different versions of Foreman.
+The required version of Foreman can be set in env variable `TEST_API_VERSION`. Make sure the requested data are in `test/unit/data/<version>/`.
+```bash
+$ TEST_API_VERSION=1.13 bundle exec rake test
+```
+
+## Testing API expectations
+
+`hammer-cli-foreman` comes with a set of helpers and extensions to [Mocha](https://github.com/freerange/mocha) library for easier testing and better error detection.
+
+The extensions are included in `test/test_helper.rb` by default. It's also possible to include it elswhere when needed:
+```ruby
+require 'hammer_cli_foreman/testing/api_expectations'
+include HammerCLIForeman::Testing::APIExpectations
+```
+
+### Available helper functions
+
+**`expects_call`**`(resource=nil, action=nil, note=nil, &block)`
+* `resource` Symbol with plural name of the resource.
+* `action` Symbol with action name.
+* `note` String with description of the expected call to make its identification easier when the expectation fails.
+* `&block` Optional block for testing parameters.
+
+ `expects_call` is often used together with `with_params` that set expectation on parameters that need to be included in the call:
+```ruby
+describe 'hostgroup create' do
+  it 'passes architecture id' do
+    api_expects(:hostgroups, :create, 'Create hostgroup with architecture_id').with_params('hostgroup' => {
```

Comment:
Why the doc describe `expects_call` and the example uses `api_expects`. If `api_expects` is recommended helper to use `expects_call` it would be less confusing to not mention the `expects_call` at all or just as a side pointer for further study.